### PR TITLE
Adding primaryContainerName implementation to podBuilder

### PR DESCRIPTION
**Pull request by niliayu**
# TL;DR
https://github.com/flyteorg/flyte/issues/2703

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [ ] Code completed
 - [ ] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
Most of the detailed discussion leading up to this change is in [this slack thread](https://flyte-org.slack.com/archives/CNMKCU6FR/p1657606862562419).

## Tracking Issue
fixes https://github.com/flyteorg/flyte/issues/2703

## Follow-up issue
_NA_

TODO (all completed):
- [x] Need to replace `'default'` name with something real
- [x] Need to refactor `sidecarPodBuilder.updatePodMetadata` to use the new naming
- [x] Test everything

### CODE REVIEW

**Comment by niliayu** [comment]
I've got this running with a pod-template, here's an example pod that was created with these changes:
<details>
<summary>Example pod</summary>
```
apiVersion: v1
kind: Pod
metadata:
  name: a4lmk5ws6qkfbhx94dxc-n0-0
  namespace: flytesnacks-development
  uid: 68e6dd69-4162-435f-ac1e-4c0da75aca07
  resourceVersion: '187993'
  creationTimestamp: '2022-08-05T21:49:22Z'
  labels:
    domain: development
    execution-id: a4lmk5ws6qkfbhx94dxc
    interruptible: 'false'
    node-id: n0
    project: flytesnacks
    shard-key: '1'
    task-name: flyte-project-example-task
    workflow-name: flyte-project-example-workflow
  annotations:
    cluster-autoscaler.kubernetes.io/safe-to-evict: 'false'
    primary_container_name: a4lmk5ws6qkfbhx94dxc-n0-0
  ownerReferences:
    - apiVersion: flyte.lyft.com/v1alpha1
      kind: flyteworkflow
      name: a4lmk5ws6qkfbhx94dxc
      uid: 9450a296-d243-4c78-a8f3-7ffc1e3e6b33
      controller: true
      blockOwnerDeletion: true
  finalizers:
    - flyte/flytek8s
  selfLink: /api/v1/namespaces/flytesnacks-development/pods/a4lmk5ws6qkfbhx94dxc-n0-0
status:
  phase: Pending
  conditions:
    - type: PodScheduled
      status: 'False'
      lastProbeTime: null
      lastTransitionTime: '2022-08-05T21:49:22Z'
      reason: Unschedulable
      message: '0/1 nodes are available: 1 Insufficient cpu.'
  qosClass: Guaranteed
spec:
  volumes:
    - name: apmsocketpath
      hostPath:
        path: /var/run/datadog/
        type: ''
    - name: kube-api-access-r486f
      projected:
        sources:
          - serviceAccountToken:
              expirationSeconds: 3607
              path: token
          - configMap:
              name: kube-root-ca.crt
              items:
                - key: ca.crt
                  path: ca.crt
          - downwardAPI:
              items:
                - path: namespace
                  fieldRef:
                    apiVersion: v1
                    fieldPath: metadata.namespace
        defaultMode: 420
  containers:
    - name: default
      image: docker.io/rwgrim/docker-noop
      args:
        - pyflyte-fast-execute
        - '--additional-distribution'
        - >-
          s3://my-s3-bucket/xs/flytesnacks/development/TDUWBF5VL7DUC4LC6UBRL5A4YE======/fast16282b6fa84bc0e633804f194037d1eb.tar.gz
        - '--dest-dir'
        - .
        - '--'
        - pyflyte-execute
        - '--inputs'
        - >-
          s3://my-s3-bucket/metadata/propeller/flytesnacks-development-a4lmk5ws6qkfbhx94dxc/n0/data/inputs.pb
        - '--output-prefix'
        - >-
          s3://my-s3-bucket/metadata/propeller/flytesnacks-development-a4lmk5ws6qkfbhx94dxc/n0/data/0
        - '--raw-output-data-prefix'
        - s3://my-s3-bucket/test/5t/a4lmk5ws6qkfbhx94dxc-n0-0
        - '--checkpoint-path'
        - s3://my-s3-bucket/test/5t/a4lmk5ws6qkfbhx94dxc-n0-0/_flytecheckpoints
        - '--prev-checkpoint'
        - '""'
        - '--resolver'
        - flytekit.core.python_auto_container.default_task_resolver
        - '--'
        - task-module
        - flyte.project.example
        - task-name
        - example_task
      env:
        - name: FLYTE_INTERNAL_EXECUTION_WORKFLOW
          value: >-
            flytesnacks:development:flyte.project.example.example_workflow
        - name: FLYTE_INTERNAL_EXECUTION_ID
          value: a4lmk5ws6qkfbhx94dxc
        - name: FLYTE_INTERNAL_EXECUTION_PROJECT
          value: flytesnacks
        - name: FLYTE_INTERNAL_EXECUTION_DOMAIN
          value: development
        - name: FLYTE_ATTEMPT_NUMBER
          value: '0'
        - name: FLYTE_INTERNAL_TASK_PROJECT
          value: flytesnacks
        - name: FLYTE_INTERNAL_TASK_DOMAIN
          value: development
        - name: FLYTE_INTERNAL_TASK_NAME
          value: flyte.project.example.example_task
        - name: FLYTE_INTERNAL_TASK_VERSION
          value: manual-register2
        - name: FLYTE_INTERNAL_PROJECT
          value: flytesnacks
        - name: FLYTE_INTERNAL_DOMAIN
          value: development
        - name: FLYTE_INTERNAL_NAME
          value: flyte.project.example.example_task
        - name: FLYTE_INTERNAL_VERSION
          value: manual-register2
        - name: FLYTE_AWS_ACCESS_KEY_ID
          value: minio
        - name: FLYTE_AWS_SECRET_ACCESS_KEY
          value: miniostorage
        - name: AWS_METADATA_SERVICE_TIMEOUT
          value: '5'
        - name: AWS_METADATA_SERVICE_NUM_ATTEMPTS
          value: '20'
        - name: FLYTE_AWS_ENDPOINT
          value: http://minio.flyte:9000
      resources:
        limits:
          cpu: '2'
          memory: 200Mi
        requests:
          cpu: '2'
          memory: 200Mi
      volumeMounts:
        - name: apmsocketpath
          mountPath: /var/run/datadog
        - name: kube-api-access-r486f
          readOnly: true
          mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      terminationMessagePath: /dev/termination-log
      terminationMessagePolicy: File
      imagePullPolicy: Always
    - name: a4lmk5ws6qkfbhx94dxc-n0-0
      image: gcr.io/pachama-labs/project/project:0.68.0
      args:
        - pyflyte-fast-execute
        - '--additional-distribution'
        - >-
          s3://my-s3-bucket/xs/flytesnacks/development/TDUWBF5VL7DUC4LC6UBRL5A4YE======/fast16282b6fa84bc0e633804f194037d1eb.tar.gz
        - '--dest-dir'
        - .
        - '--'
        - pyflyte-execute
        - '--inputs'
        - >-
          s3://my-s3-bucket/metadata/propeller/flytesnacks-development-a4lmk5ws6qkfbhx94dxc/n0/data/inputs.pb
        - '--output-prefix'
        - >-
          s3://my-s3-bucket/metadata/propeller/flytesnacks-development-a4lmk5ws6qkfbhx94dxc/n0/data/0
        - '--raw-output-data-prefix'
        - s3://my-s3-bucket/test/5t/a4lmk5ws6qkfbhx94dxc-n0-0
        - '--checkpoint-path'
        - s3://my-s3-bucket/test/5t/a4lmk5ws6qkfbhx94dxc-n0-0/_flytecheckpoints
        - '--prev-checkpoint'
        - '""'
        - '--resolver'
        - flytekit.core.python_auto_container.default_task_resolver
        - '--'
        - task-module
        - flyte.project.example
        - task-name
        - example_task
        - pyflyte-fast-execute
        - '--additional-distribution'
        - >-
          s3://my-s3-bucket/xs/flytesnacks/development/TDUWBF5VL7DUC4LC6UBRL5A4YE======/fast16282b6fa84bc0e633804f194037d1eb.tar.gz
        - '--dest-dir'
        - .
        - '--'
        - pyflyte-execute
        - '--inputs'
        - >-
          s3://my-s3-bucket/metadata/propeller/flytesnacks-development-a4lmk5ws6qkfbhx94dxc/n0/data/inputs.pb
        - '--output-prefix'
        - >-
          s3://my-s3-bucket/metadata/propeller/flytesnacks-development-a4lmk5ws6qkfbhx94dxc/n0/data/0
        - '--raw-output-data-prefix'
        - s3://my-s3-bucket/test/5t/a4lmk5ws6qkfbhx94dxc-n0-0
        - '--checkpoint-path'
        - s3://my-s3-bucket/test/5t/a4lmk5ws6qkfbhx94dxc-n0-0/_flytecheckpoints
        - '--prev-checkpoint'
        - '""'
        - '--resolver'
        - flytekit.core.python_auto_container.default_task_resolver
        - '--'
        - task-module
        - flyte.project.example
        - task-name
        - example_task
      env:
        - name: FLYTE_INTERNAL_EXECUTION_WORKFLOW
          value: >-
            flytesnacks:development:flyte.project.example.example_workflow
        - name: FLYTE_INTERNAL_EXECUTION_ID
          value: a4lmk5ws6qkfbhx94dxc
        - name: FLYTE_INTERNAL_EXECUTION_PROJECT
          value: flytesnacks
        - name: FLYTE_INTERNAL_EXECUTION_DOMAIN
          value: development
        - name: FLYTE_ATTEMPT_NUMBER
          value: '0'
        - name: FLYTE_INTERNAL_TASK_PROJECT
          value: flytesnacks
        - name: FLYTE_INTERNAL_TASK_DOMAIN
          value: development
        - name: FLYTE_INTERNAL_TASK_NAME
          value: flyte.project.example.example_task
        - name: FLYTE_INTERNAL_TASK_VERSION
          value: manual-register2
        - name: FLYTE_INTERNAL_PROJECT
          value: flytesnacks
        - name: FLYTE_INTERNAL_DOMAIN
          value: development
        - name: FLYTE_INTERNAL_NAME
          value: flyte.project.example.example_task
        - name: FLYTE_INTERNAL_VERSION
          value: manual-register2
        - name: FLYTE_AWS_ACCESS_KEY_ID
          value: minio
        - name: FLYTE_AWS_SECRET_ACCESS_KEY
          value: miniostorage
        - name: AWS_METADATA_SERVICE_TIMEOUT
          value: '5'
        - name: AWS_METADATA_SERVICE_NUM_ATTEMPTS
          value: '20'
        - name: FLYTE_AWS_ENDPOINT
          value: http://minio.flyte:9000
        - name: FLYTE_INTERNAL_EXECUTION_WORKFLOW
          value: >-
            flytesnacks:development:flyte.project.example.example_workflow
        - name: FLYTE_INTERNAL_EXECUTION_ID
          value: a4lmk5ws6qkfbhx94dxc
        - name: FLYTE_INTERNAL_EXECUTION_PROJECT
          value: flytesnacks
        - name: FLYTE_INTERNAL_EXECUTION_DOMAIN
          value: development
        - name: FLYTE_ATTEMPT_NUMBER
          value: '0'
        - name: FLYTE_INTERNAL_TASK_PROJECT
          value: flytesnacks
        - name: FLYTE_INTERNAL_TASK_DOMAIN
          value: development
        - name: FLYTE_INTERNAL_TASK_NAME
          value: flyte.project.example.example_task
        - name: FLYTE_INTERNAL_TASK_VERSION
          value: manual-register2
        - name: FLYTE_INTERNAL_PROJECT
          value: flytesnacks
        - name: FLYTE_INTERNAL_DOMAIN
          value: development
        - name: FLYTE_INTERNAL_NAME
          value: flyte.project.example.example_task
        - name: FLYTE_INTERNAL_VERSION
          value: manual-register2
        - name: FLYTE_AWS_ACCESS_KEY_ID
          value: minio
        - name: FLYTE_AWS_SECRET_ACCESS_KEY
          value: miniostorage
        - name: AWS_METADATA_SERVICE_TIMEOUT
          value: '5'
        - name: AWS_METADATA_SERVICE_NUM_ATTEMPTS
          value: '20'
        - name: FLYTE_AWS_ENDPOINT
          value: http://minio.flyte:9000
      resources:
        limits:
          cpu: '2'
          memory: 200Mi
        requests:
          cpu: '2'
          memory: 200Mi
      volumeMounts:
        - name: kube-api-access-r486f
          readOnly: true
          mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      terminationMessagePath: /dev/termination-log
      terminationMessagePolicy: FallbackToLogsOnError
      imagePullPolicy: IfNotPresent
  restartPolicy: Never
  terminationGracePeriodSeconds: 30
  dnsPolicy: ClusterFirst
  serviceAccountName: default
  serviceAccount: default
  securityContext: {}
  schedulerName: default-scheduler
  tolerations:
    - key: node.kubernetes.io/not-ready
      operator: Exists
      effect: NoExecute
      tolerationSeconds: 300
    - key: node.kubernetes.io/unreachable
      operator: Exists
      effect: NoExecute
      tolerationSeconds: 300
  priority: 0
  enableServiceLinks: true
  preemptionPolicy: PreemptLowerPriority
```
</details>
Next up I need to figure out a test workflow for sidecars, since my understanding of how those should look is still a bit high level, and I'm not very sure I've changed the sidecar code the way we want or not. Once I get that sorted I'll have a look at unit tests I should add for this.

File: go/tasks/pluginmachinery/flytek8s/pod_helper.go [hunk 1]
```diff
@@ -135,7 +136,11 @@ func ToK8sPodSpec(ctx context.Context, tCtx pluginsCore.TaskExecutionContext) (*
 	return pod, nil
 }
 
-func BuildPodWithSpec(podTemplate *v1.PodTemplate, podSpec *v1.PodSpec) (*v1.Pod, error) {
+func PrimaryContainerNameMatches(containerName string, templateContainerName string, primaryContainerName string) bool {
```
Comment (hamersaw) [inline go/tasks/pluginmachinery/flytek8s/pod_helper.go:139]: I don't think this necessarily needs to be a separate function. Probably, OK to just use inline.

File: go/tasks/pluginmachinery/flytek8s/pod_helper.go [hunk 2]
```diff
@@ -150,8 +155,22 @@ func BuildPodWithSpec(podTemplate *v1.PodTemplate, podSpec *v1.PodSpec) (*v1.Pod
...
+					err := mergo.Merge(newContainer, container)
+					if err != nil {
+						return nil, err
+					}
+					mergedContainers = append(mergedContainers, *newContainer)
+				} else {
+					mergedContainers = append(mergedContainers, container)
+				}
+			}
+		}
```
Comment (hamersaw) [inline go/tasks/pluginmachinery/flytek8s/pod_helper.go:172]: May need to double-check the loop and container logic here. I think in this scenario the case where the template containers:
```
["foo", "bar", "default"]
```
and the podSpec from flyte contains containers:
```
["baz", "default"]
```
the resulting container set will be
```
["foo", "bar", "default"]
```
when it should be:
```
["baz", "default"]
```
with the configuration of default using the podTemplate configuration. Am I thinking about this correctly? Or should the result be:
```
["foo", "bar", "baz", "default"]
```
where any container in the template is either the base configuration for a pod container or added?

**Review by hamersaw — COMMENTED** [review]
Looks great so far! Left a few comments, I think you'll have to fix a few things in the sidecar plugin once you get to it. [Here](https://docs.flyte.org/projects/cookbook/en/latest/auto/integrations/kubernetes/pod/pod.html#sphx-glr-auto-integrations-kubernetes-pod-pod-py) is documentation on defining a task to use the sidecar plugin.

File: go/tasks/pluginmachinery/flytek8s/pod_helper.go (hunk 2 above)
Comment (niliayu) [reply go/tasks/pluginmachinery/flytek8s/pod_helper.go:172]: In the behavior that you were expecting, do `foo` and `bar` from the template also get merged into the primary container `baz`? Or the default? (I think I don't have a good example to think through for why we'd want this behavior, since the use case I'm concerned with is more limited)

Comment (hamersaw) [reply go/tasks/pluginmachinery/flytek8s/pod_helper.go:172]: Yeah, so I don't think we should ever merge anything other than the "default" or "primary-container" in the template with the default podSepc container.

In this example I think the "foo" and "bar" containers from the template will be added to the podSpec, without merging with anything. The only use-case I can think of for this functionality is the ability to add a container to every pod that is started by Flyte. For example if we had the PodTemplate:
```
apiVersion: v1
kind: PodTemplate
metadata:
  name: flyte-default-template
  namespace: flyte
template:
  spec:
    containers:
      - name: primary-container-name
        volumeMounts:
          - name: foo
            mountPath: /foo
      - name: noop
        image: docker.io/rwgrim/docker-noop
```
We would obviously use the `primary-container-name` container as the base configuration for all of the primary containers. The `noop` container, while admittedly non-useful in this example, could be something more interesting. The question is, should we either:
(1) never use this container
(2) add the container to the Pod

I think for simplicity option 1 is better, I think we would have to change logic of how the container plugin executes with option 2. Also, this is something we can revisit in the future if we want.

Comment (niliayu) [reply go/tasks/pluginmachinery/flytek8s/pod_helper.go:172]: Got it thank you, that makes a ton of sense! I was definitely thinking of option 2 because I wasn't sure if we would be okay with not running it, but as I said I can't think of a clear use case for adding multiple to the pod like this, so I'm good with option 1 as it also satisfies what I'm looking for with this change!

**Review by hamersaw — APPROVED** [review]
